src/js/BookReader/BookModel.js

```diff
@@ -0,0 +1,286 @@
+// @ts-check
+import { clamp } from './utils.js';
+/** @typedef {import('./options.js').PageData} PageData */
+/** @typedef {import('../BookReader.js').default} BookReader */
+
+/**
+ * Contains information about the Book/Document independent of the way it is
+ * being rendering. Nothing here should reference e.g. the mode, zoom, etc.
+ * It's just information about the book and its pages (usually as specified
+ * in the BookReader data option.)
+ */
+export class BookModel {
+  /**
+   * @param {BookReader} br 
+   */
+  constructor(br) {
+    this.br = br;
+
+    /** @type {{width: number, height: number}} memoize storage */
+    this._medianPageSize = null;
+    /** @type {[PageData[], number]} */
+    this._getDataFlattenedCached = null;
+  }
+
+  /**
+   * Memoized
+   * @return {{width: number, height: number}}
+   */
+  getMedianPageSize() {
+    if (this._medianPageSize) {
+      return this._medianPageSize;
+    }
+  
+    // A little expensive but we just do it once
+    const widths = [];
+    const heights = [];
+    for (let i = 0; i < this.getNumLeafs(); i++) {
+      widths.push(this.getPageWidth(i));
+      heights.push(this.getPageHeight(i));
+    }
+  
+    widths.sort();
+    heights.sort();
+  
+    this._medianPageSize = {
+      width: widths[Math.floor(widths.length / 2)],
+      height: heights[Math.floor(heights.length / 2)]
+    };
+    return this._medianPageSize;
+  }
+
+  /**
+   * Returns the page width for the given index, or first or last page if out of range
+   * @deprecated see getPageWidth
+   * @param {number} index
```

Comment:
This had `@deprecated` before I got here; I just copied it over. It's used in a _lot_ of places to be deprecated though :/ Punting on adding `@return` here, since the system can deduce the type here (and it's really easy to find :) )